cli/azd/pkg/auth/errors.go

```diff
@@ -0,0 +1,131 @@
+package auth
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+
+	msal "github.com/AzureAD/microsoft-authentication-library-for-go/apps/errors"
+)
+
+const authFailedPrefix string = "failed to authenticate"
+
+// An error response from Azure Active Directory.
+//
+// See https://www.rfc-editor.org/rfc/rfc6749#section-5.2 for OAuth 2.0 spec
+// See https://learn.microsoft.com/en-us/azure/active-directory/develop/reference-aadsts-error-codes for AAD error codes
+type AadErrorResponse struct {
+	Error            string `json:"error"`
+	ErrorDescription string `json:"error_description"`
+	ErrorCodes       []int  `json:"error_codes"`
+	Timestamp        string `json:"timestamp"`
+	TraceId          string `json:"trace_id"`
+	CorrelationId    string `json:"correlation_id"`
+	ErrorUri         string `json:"error_uri"`
+}
+
+// AuthFailedError indicates an authentication request has failed.
+type AuthFailedError struct {
```

Comment:
Just wondering - do you think we should harmonize this with ErrNoCurrentUser?  I could imagine that we could augment this error type such that errors.Is(<this>, auth.ErrNoCurrentUser) return, at least in some cases (maybe you want to make it conditional the `invalid_grant` or `interaction_required` error codes?.  I'm not sure if that would help us at all.

I could also see us just deleting ErrNoCurrentUser?  I'm just checking to make sure leaving it around isn't going to be leaving ourselves a footgun.

Reply:
ErrNoCurrentUser is currently returned when we know the user isn't logged in, i.e:
-  `auth.json` isn't configured
- The token cache is missing. This ends up hitting the default return [here](https://github.com/Azure/azure-dev/blob/ad079ad6bfd70874d022ae533c9b70849e9ac091/cli/azd/pkg/auth/manager.go#L243) because public.Accounts is nil

The purpose of the error is when we failed to use the stored ID tokens to retrieve an access token. Let me think about how they could logically be related better.

Reply:
Gave the idea the thought it deserved.

After thinking through the problem, I've made some changes:
- AuthFailedError is now the generic error wrapper around MSAL related errors related to token exchange, which is heavily inspired for `azidentity`. I'm not sure if MSAL would ever go into the direction of providing parsed error responses (it makes error handling more complicated as a general library). This gives us the full flexibility of being able to examine error codes, make error messages better, log errors -- which MSAL doesn't yet provide functionality for.
- ReloginRequiredError has been added as a separate type. This allows for examining the parsed response, and determine if the appropriate error is a handled one. This could be useful in the future if we want to start on items like #2184 

`ErrNoCurrentUser` is important as an error handling condition when the user simply isn't logged in. It should be kept.

I hope this is still in-line with YAGNI / golang philosophy. I definitely think the error adapter is powerful, and perhaps we can think about separating out the library from app pieces if needed at some point.

The change ended up being a little bigger than the original, so I'd be happy if you want to give it another pass, @ellismg .